Modify rule S3457: LaYC on printf errors (#3106)
SonarSource/rspec: rules/S3457/cfamily/rule.adoc

== Why is this an issue?

> Because `printf` format strings are interpreted at runtime, rather than validated by the compiler, they can contain errors that result in the wrong string being created. This rule statically validates the correlation of `printf` format strings and their arguments.

Reviewer: ```suggestion
Because `printf` format strings are interpreted at runtime rather than validated by the compiler, they can contain errors that result in the wrong string being created. This rule statically validates the correlation of `printf` format strings and their arguments.
```

[source,cpp]
----

> printf("%d", 1, 2); // Noncompliant; the second argument "2" is unused
> printf("%0-f", 1.2); // Noncompliant; flag "0" is ignored because of "-"

Reviewer: ```suggestion
printf("%d", 1, 2); // Noncompliant: the second argument "2" is unused
printf("%0-f", 1.2); // Noncompliant: flag "0" is ignored because of "-"
```
see https://sonarsource.slack.com/archives/C04P0SB2KJP/p1686055543993659
(could not find anything in the docs though)

----

> Starting with C++20, std::format should be preferred, as it is more readable and validated at compile-time, making it more secure. S6494 covers that.

Reviewer: ```suggestion
Starting with C++20, `std::format` should be preferred, as it is more readable and validated at compile-time, making it more secure. S6494 covers that.
```

> The related rule S2275 is about errors that will create undefined behavior, while this rule is about errors that produce an unexpected string.

Reviewer: Nitpick, but I would begin the sentence with the current rule as it is the one that interests more the user:
```suggestion
This rule is about errors that produce an unexpected string. For errors that will create undefined behavior, see the related rule S2275.
```

=== Exceptions

> This rule will only work if the format string is provided as a string literal.

Reviewer: The exception sentence does not look like an exception but a limitation.
I would either remove the exception paragraph (and include its content in the description) or reword the sentence "This rule does not raise issues when ..."
(I prefer the first option as it really looks like a technical limitation)

Author: Very true. I like this exception/limitation dichotomy.

> == Resources

Reviewer: I would add a link to std::format, e.g. https://en.cppreference.com/w/cpp/utility/format/format (not very user-friendly, but we'll add the blog post when we have one about std::format)
